RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc

```diff
@@ -7,33 +13,218 @@ namespace {
+  _recHitsEE = cc.consumes<EcalRecHitCollection>(conf.getParameter<edm::InputTag>("recHitsEELabel"));
+  _vertices  = cc.consumes<reco::VertexCollection>(conf.getParameter<edm::InputTag>("verticesLabel"));
+  
+  autoDetectBunchSpacing_ = conf.getParameter<bool>("autoDetectBunchSpacing");
+
+  if (autoDetectBunchSpacing_) {
+    bunchSpacing_ = cc.consumes<int>(edm::InputTag("addPileupInfo","bunchSpacing"));
+    bunchSpacingManual_ = 0;
+  }
+  else {
+    bunchSpacingManual_ = conf.getParameter<int>("bunchSpacing");
+  }
+  
+  _condnames_mean_50ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize1_mean_50ns");
+  _condnames_mean_50ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize2_mean_50ns");
+  _condnames_mean_50ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize3_mean_50ns");
+  _condnames_mean_50ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize1_mean_50ns");
+  _condnames_mean_50ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize2_mean_50ns");
+  _condnames_mean_50ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize3_mean_50ns");
+
+  _condnames_sigma_50ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize1_sigma_50ns");
+  _condnames_sigma_50ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize2_sigma_50ns");
+  _condnames_sigma_50ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize3_sigma_50ns");
+  _condnames_sigma_50ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize1_sigma_50ns");
+  _condnames_sigma_50ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize2_sigma_50ns");
+  _condnames_sigma_50ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize3_sigma_50ns");
+  
+  _condnames_mean_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize1_mean_25ns");
+  _condnames_mean_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize2_mean_25ns");
+  _condnames_mean_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize3_mean_25ns");
+  _condnames_mean_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize1_mean_25ns");
+  _condnames_mean_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize2_mean_25ns");
+  _condnames_mean_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize3_mean_25ns");
+
+  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize1_sigma_25ns");
+  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize2_sigma_25ns");
+  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize3_sigma_25ns");
+  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize1_sigma_25ns");
+  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize2_sigma_25ns");
+  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize3_sigma_25ns");  
+  
+}
+
+
+void PFClusterEMEnergyCorrector::correctEnergies(const edm::Event &evt, const edm::EventSetup &es, const reco::PFCluster::EEtoPSAssociation &assoc, reco::PFClusterCollection& cs) {
+
+  int bunchspacing = 450;  
+  
+  if (autoDetectBunchSpacing_) {
```

Comment:
One of the things that tripped us up last time is that the proper solution
should eventually have this information produced as a run product rather
than an event product.  This would/does require larger changes to the
mixing module to accomplish the current mc-based detection.

If we're happy living with this as an event product for now then moving it
to a single producer could indeed work.

On 5 February 2015 at 15:36, David Lange <notifications@github.com> wrote:

> In RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc
> <https://github.com/cms-sw/cmssw/pull/7558#discussion_r24166456>:
>
> > +
> > +  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize1_sigma_25ns");
> > +  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize2_sigma_25ns");
> > +  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize3_sigma_25ns");
> > +  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize1_sigma_25ns");
> > +  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize2_sigma_25ns");
> > +  _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize3_sigma_25ns");
> > +
> > +}
> > +
> > +
> > +void PFClusterEMEnergyCorrector::correctEnergies(const edm::Event &evt, const edm::EventSetup &es, const reco::PFCluster::EEtoPSAssociation &assoc, reco::PFClusterCollection& cs) {
> > +
> > +  int bunchspacing = 450;
> > +
> > +  if (autoDetectBunchSpacing_) {
>
> regarding a "proper" solution.. one option we can easily do is to at least
> centralize this "bad" solution into a single edproducer. Then if we get to
> the point that we can derive the bunch spacing from the conditions, we have
> to update one place. Reasons that is not a good way to proceed?
> … <#14b5a2c16104590b_>
> On Feb 4, 2015, at 5:52 PM, Josh Bendavid <notifications@github.com>
> wrote: In
> RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc: > + >
> +
> _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize1_sigma_25ns");
> > +
> _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize2_sigma_25ns");
> > +
> _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EB_pfSize3_sigma_25ns");
> > +
> _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize1_sigma_25ns");
> > +
> _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize2_sigma_25ns");
> > +
> _condnames_sigma_25ns.push_back("GBRForestD_ecalPFClusterCor_EE_pfSize3_sigma_25ns");
> > + > +} > + > + > +void PFClusterEMEnergyCorrector::correctEnergies(const
> edm::Event &evt, const edm::EventSetup &es, const
> reco::PFCluster::EEtoPSAssociation &assoc, reco::PFClusterCollection& cs) {
> > + > + int bunchspacing = 450; > + > + if (autoDetectBunchSpacing_) { My
> hope is that making this increasingly horrible leads to a proper solution
> being implemented sooner... � Reply to this email directly or view it on
> GitHub.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/7558/files#r24166456>.
>